File: emulinker/src/main/java/org/emulinker/kaillera/model/impl/AutoFireScanner2.java
package org.emulinker.kaillera.model.impl;

import java.util.*;
import java.util.concurrent.*;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.emulinker.kaillera.model.*;
import org.emulinker.util.*;

public class AutoFireScanner2 implements AutoFireDetector {
  protected static Log log = LogFactory.getLog(AutoFireScanner2.class);

  // MAX DELAY, MIN REPEITIONS
  private static int SENSITIVITY_TABLE[][] = {
    {0, 0}, // 0 means disable autofire detect
    {2, 13}, // 1 is least sensitive
    {3, 11},
    {4, 9},
    {5, 7},
    {6, 5} // 5 is most sensitive
  };

  protected static ExecutorService executor = Executors.newCachedThreadPool();

  protected KailleraGame game;
  protected int sensitivity;
  protected int maxDelay;
  protected int minReps;

  protected ScanningJob[] scanningJobs;

  public AutoFireScanner2(KailleraGame game, int sensitivity) {
    this.game = game;
    setSensitivity(sensitivity);
  }

  @Override
  public int getSensitivity() {
    return sensitivity;
  }

  @Override
  public void setSensitivity(int sensitivity) {
    if (sensitivity < 0 || sensitivity > 5) this.sensitivity = 0;
    else {
      this.sensitivity = sensitivity;
      maxDelay = SENSITIVITY_TABLE[sensitivity][0];
      minReps = SENSITIVITY_TABLE[sensitivity][1];
    }
  }

  @Override
  public void start(int numPlayers) {
    if (sensitivity <= 0) return;

    scanningJobs = new ScanningJob[numPlayers];
  }

  @Override
  public void addPlayer(KailleraUser player, int playerNumber) {
    if (sensitivity <= 0 || scanningJobs == null) return;

    scanningJobs[(playerNumber - 1)] = new ScanningJob(player, playerNumber);
  }

  @Override
  public void stop(int playerNumber) {
    if (sensitivity <= 0 || scanningJobs == null) return;

    scanningJobs[(playerNumber - 1)].stop();
  }

  @Override
  public void stop() {
    if (sensitivity <= 0 || scanningJobs == null) return;

    for (int i = 0; i < scanningJobs.length; i++) scanningJobs[i].stop();
  }

  @Override
  public void addData(int playerNumber, byte[] data, int bytesPerAction) {
    if (sensitivity <= 0 || scanningJobs == null) return;

    scanningJobs[(playerNumber - 1)].addData(data, bytesPerAction);
  }

  protected class ScanningJob implements Runnable {
    private KailleraUser user;
    private int playerNumber;

    private int bytesPerAction = -1;
    private int sizeLimit;
    private int bufferSize = 5;
    private int size = 0;

    private byte[][] buffer;
    private int head = 0;
    private int tail = 0;
    private int pos = 0;

    private boolean running = false;
    private boolean stopFlag = false;

    protected ScanningJob(KailleraUser user, int playerNumber) {
      this.user = user;
      this.playerNumber = playerNumber;

      sizeLimit = ((maxDelay + 1) * minReps * 5);
      buffer = new byte[bufferSize][sizeLimit];
    }

    protected synchronized void addData(byte[] data, int bytesPerAction) {
      if ((pos + data.length) >= sizeLimit) {
        int firstSize = (sizeLimit - pos);
        //				log.debug("firstSize="+firstSize);
        System.arraycopy(data, 0, buffer[tail], pos, firstSize);
        // tail = ((tail + 1) % bufferSize);
        tail++;
        if (tail == bufferSize) tail = 0;
        //				log.debug("tail="+tail);
        System.arraycopy(data, firstSize, buffer[tail], 0, (data.length - firstSize));
        pos = (data.length - firstSize);
        //				log.debug("pos="+pos);
        size++;

        if (this.bytesPerAction <= 0) this.bytesPerAction = bytesPerAction;

        if (!running) executor.submit(this);
      } else {
        System.arraycopy(data, 0, buffer[tail], pos, data.length);
        pos += data.length;
        //				log.debug("pos="+pos);
      }
    }

    protected void stop() {
      this.stopFlag = true;
    }

    @Override
    public void run() {
      //			long st = System.currentTimeMillis();
      synchronized (this) {
        running = true;
      }

      try {
        while (size > 0 && !stopFlag) {
          byte[] data = null;
          synchronized (this) {
            data = buffer[head];
            //						log.debug("Scanning " + data.length + " bytes from buffer position " + head);
            // head = ((head+1) % bufferSize);
            head++;
            if (head == bufferSize) head = 0;

            size--;
          }

          // determine the number of actions in this array
          int actionCount = (data.length / bytesPerAction);
          byte[] thisAction = new byte[bytesPerAction];
          byte[] lastAction = new byte[bytesPerAction];
          byte[] actionA = new byte[bytesPerAction];
          int aPos = 0;
          int aCount = 0;
          int aSequence = 0;
          int lastASequence = 0;
          int aSequenceCount = 0;
          byte[] actionB = new byte[bytesPerAction];
          int bPos = 0;
          int bCount = 0;
          int bSequence = 0;
          int lastBSequence = 0;
          int bSequenceCount = 0;

          for (int i = 0; i < actionCount; i++) {
            System.arraycopy(data, (i * bytesPerAction), thisAction, 0, bytesPerAction);
            //						log.debug("thisAction=" + EmuUtil.bytesToHex(thisAction) + " actionA=" +
            // EmuUtil.bytesToHex(actionA) + " aCount=" + aCount + " actionB=" +
            // EmuUtil.bytesToHex(actionB) + " bCount=" + bCount + " aSequence=" + aSequence + "
            // aSequenceCount=" + aSequenceCount + " bSequence=" + bSequence + " bSequenceCount=" +
            // bSequenceCount);

            if (aCount == 0) {
              System.arraycopy(thisAction, 0, actionA, 0, bytesPerAction);
              aPos = i;
              aCount = 1;
              aSequence = 1;
            } else if (Arrays.equals(thisAction, actionA)) {
              aCount++;
              if (Arrays.equals(thisAction, lastAction)) aSequence++;
              else {
                if (lastASequence == aSequence && aSequence <= maxDelay) aSequenceCount++;
                else aSequenceCount = 0;
                lastASequence = aSequence;
                aSequence = 1;
              }
            } else if (bCount == 0) {
              System.arraycopy(thisAction, 0, actionB, 0, bytesPerAction);
              bPos = i;
              bCount = 1;
              bSequence = 1;
            } else if (Arrays.equals(thisAction, actionB)) {
              bCount++;
              if (Arrays.equals(thisAction, lastAction)) bSequence++;
              else {
                if (lastBSequence == bSequence && bSequence <= maxDelay) bSequenceCount++;
                else bSequenceCount = 0;
                lastBSequence = bSequence;
                bSequence = 1;
              }
            } else {
              actionA = lastAction;
              aCount = 1;
              aSequence = 1;
              aSequenceCount = 0;
              actionB = thisAction;
              bCount = 1;
              bSequence = 0;
              bSequenceCount = 0;
            }

            System.arraycopy(thisAction, 0, lastAction, 0, bytesPerAction);

            //						if(aSequenceCount >= 3 && bSequenceCount >= 3 && !stopFlag)
            //						{
            //							log.debug("thisAction=" + EmuUtil.bytesToHex(thisAction) + " actionA=" +
            // EmuUtil.bytesToHex(actionA) + " aCount=" + aCount + " actionB=" +
            // EmuUtil.bytesToHex(actionB) + " bCount=" + bCount + " aSequence=" + aSequence + "
            // aSequenceCount=" + aSequenceCount + " bSequence=" + bSequence + " bSequenceCount=" +
            // bSequenceCount);
            //						}

            if (aSequenceCount >= minReps && bSequenceCount >= minReps && !stopFlag) {
              KailleraGameImpl gameImpl = (KailleraGameImpl) game;
              gameImpl.announce(
                  EmuLang.getString("AutoFireScanner2.AutoFireDetected", user.getName()),
                  null); //$NON-NLS-1$
              log.info(
                  "AUTOUSERDUMP\t"
                      + EmuUtil.DATE_FORMAT.format(gameImpl.getStartDate())
                      + "\t"
                      + (aSequence < bSequence ? aSequence : bSequence)
                      + "\t"
                      + game.getID()
                      + "\t"
                      + game.getRomName()
                      + "\t"
                      + user.getName()
                      + "\t"
                      + user.getSocketAddress()
                          .getAddress()
                          .getHostAddress()); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
              // //$NON-NLS-4$ //$NON-NLS-5$ //$NON-NLS-6$
              //							log.debug("thisAction=" + EmuUtil.bytesToHex(thisAction) + " actionA=" +
              // EmuUtil.bytesToHex(actionA) + " aCount=" + aCount + " actionB=" +
              // EmuUtil.bytesToHex(actionB) + " bCount=" + bCount + " aSequence=" + aSequence + "
              // aSequenceCount=" + aSequenceCount + " bSequence=" + bSequence + " bSequenceCount="
              // + bSequenceCount);
              break;
            }
          }
        }
      } catch (Exception e) {
        log.error(
            "AutoFireScanner2 thread for " + user + " caught exception!",
            e); //$NON-NLS-1$ //$NON-NLS-2$
      } finally {
        synchronized (this) {
          running = false;
        }
      }

      //			long et = (System.currentTimeMillis()-st);
      //			log.debug("Scanning completed in " + et + " ms");
    }
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/model/impl/AutoFireScanner2.kt
package org.emulinker.kaillera.model.impl

import com.google.common.flogger.FluentLogger
import java.lang.Exception
import java.lang.Runnable
import java.util.Arrays
import java.util.concurrent.Executors
import org.emulinker.kaillera.model.KailleraGame
import org.emulinker.kaillera.model.KailleraUser
import org.emulinker.util.EmuLang.getString
import org.emulinker.util.EmuUtil

private val logger = FluentLogger.forEnclosingClass()

class AutoFireScanner2(private var game: KailleraGame, sensitivity: Int) : AutoFireDetector {
  override var sensitivity = 0
    set(value) {
      if (value in 0..5) {
        field = value
        maxDelay = SENSITIVITY_TABLE[value][0]
        minReps = SENSITIVITY_TABLE[value][1]
      } else {
        field = 0
      }
    }

  protected var maxDelay = 0
  protected var minReps = 0

  protected var scanningJobs: Array<ScanningJob?>? = null

  override fun start(numPlayers: Int) {
    if (sensitivity <= 0) return
    scanningJobs = arrayOfNulls(numPlayers)
  }

  override fun addPlayer(player: KailleraUser?, playerNumber: Int) {
    if (sensitivity <= 0 || scanningJobs == null) return
    scanningJobs!![playerNumber - 1] = ScanningJob(player, playerNumber)
  }

  override fun stop(playerNumber: Int) {
    if (sensitivity <= 0 || scanningJobs == null) return
    scanningJobs!![playerNumber - 1]!!.stop()
  }

  override fun stop() {
    if (sensitivity <= 0 || scanningJobs == null) return
    for (i in scanningJobs!!.indices) scanningJobs!![i]!!.stop()
  }

  override fun addData(playerNumber: Int, data: ByteArray?, bytesPerAction: Int) {
    if (sensitivity <= 0 || scanningJobs == null) return
    scanningJobs!![playerNumber - 1]!!.addData(data, bytesPerAction)
  }

  protected inner class ScanningJob(
      private val user: KailleraUser?, private val playerNumber: Int
  ) : Runnable {
    private var bytesPerAction = -1
    private val sizeLimit: Int
    private val bufferSize = 5
    private var size = 0
    private val buffer: Array<ByteArray>
    private var head = 0
    private var tail = 0
    private var pos = 0
    private var running = false
    private var stopFlag = false
    @Synchronized
    fun addData(data: ByteArray?, bytesPerAction: Int) {
      if (pos + data!!.size >= sizeLimit) {
        val firstSize = sizeLimit - pos
        //				logger.atFine().log("firstSize="+firstSize);
        System.arraycopy(data, 0, buffer[tail], pos, firstSize)
        // tail = ((tail + 1) % bufferSize);
        tail++
        if (tail == bufferSize) tail = 0
        //				logger.atFine().log("tail="+tail);
        System.arraycopy(data, firstSize, buffer[tail], 0, data.size - firstSize)
        pos = data.size - firstSize
        //				logger.atFine().log("pos="+pos);
        size++
        if (this.bytesPerAction <= 0) this.bytesPerAction = bytesPerAction
        if (!running) executor.submit(this)
      } else {
        System.arraycopy(data, 0, buffer[tail], pos, data.size)
        pos += data.size
        //				logger.atFine().log("pos="+pos);
      }
    }

    fun stop() {
      stopFlag = true
    }

    override fun run() {
      //			long st = System.currentTimeMillis();
      synchronized(this) { running = true }
      try {
        while (size > 0 && !stopFlag) {
          var data: ByteArray? = null
          synchronized(this) {
            data = buffer[head]
            //						logger.atFine().log("Scanning " + data.length + " bytes from buffer position " +
            // head);
            // head = ((head+1) % bufferSize);
            head++
            if (head == bufferSize) head = 0
            size--
          }

          // determine the number of actions in this array
          val actionCount = data!!.size / bytesPerAction
          val thisAction = ByteArray(bytesPerAction)
          val lastAction = ByteArray(bytesPerAction)
          var actionA = ByteArray(bytesPerAction)
          var aPos = 0
          var aCount = 0
          var aSequence = 0
          var lastASequence = 0
          var aSequenceCount = 0
          var actionB = ByteArray(bytesPerAction)
          var bPos = 0
          var bCount = 0
          var bSequence = 0
          var lastBSequence = 0
          var bSequenceCount = 0
          for (i in 0 until actionCount) {
            System.arraycopy(data, i * bytesPerAction, thisAction, 0, bytesPerAction)
            //						logger.atFine().log("thisAction=" + EmuUtil.bytesToHex(thisAction) + " actionA="
            // +
            // EmuUtil.bytesToHex(actionA) + " aCount=" + aCount + " actionB=" +
            // EmuUtil.bytesToHex(actionB) + " bCount=" + bCount + " aSequence=" + aSequence + "
            // aSequenceCount=" + aSequenceCount + " bSequence=" + bSequence + " bSequenceCount=" +
            // bSequenceCount);
            if (aCount == 0) {
              System.arraycopy(thisAction, 0, actionA, 0, bytesPerAction)
              aPos = i
              aCount = 1
              aSequence = 1
            } else if (Arrays.equals(thisAction, actionA)) {
              aCount++
              if (Arrays.equals(thisAction, lastAction)) aSequence++
              else {
                if (lastASequence == aSequence && aSequence <= maxDelay) aSequenceCount++
                else aSequenceCount = 0
                lastASequence = aSequence
                aSequence = 1
              }
            } else if (bCount == 0) {
              System.arraycopy(thisAction, 0, actionB, 0, bytesPerAction)
              bPos = i
              bCount = 1
              bSequence = 1
            } else if (Arrays.equals(thisAction, actionB)) {
              bCount++
              if (Arrays.equals(thisAction, lastAction)) bSequence++
              else {
                if (lastBSequence == bSequence && bSequence <= maxDelay) bSequenceCount++
                else bSequenceCount = 0
                lastBSequence = bSequence
                bSequence = 1
              }
            } else {
              actionA = lastAction
              aCount = 1
              aSequence = 1
              aSequenceCount = 0
              actionB = thisAction
              bCount = 1
              bSequence = 0
              bSequenceCount = 0
            }
            System.arraycopy(thisAction, 0, lastAction, 0, bytesPerAction)

            //						if(aSequenceCount >= 3 && bSequenceCount >= 3 && !stopFlag)
            //						{
            //							logger.atFine().log("thisAction=" + EmuUtil.bytesToHex(thisAction) + "
            // actionA=" +
            // EmuUtil.bytesToHex(actionA) + " aCount=" + aCount + " actionB=" +
            // EmuUtil.bytesToHex(actionB) + " bCount=" + bCount + " aSequence=" + aSequence + "
            // aSequenceCount=" + aSequenceCount + " bSequence=" + bSequence + " bSequenceCount=" +
            // bSequenceCount);
            //						}
            if (aSequenceCount >= minReps && bSequenceCount >= minReps && !stopFlag) {
              val gameImpl = game as KailleraGameImpl
              gameImpl.announce(
                  getString("AutoFireScanner2.AutoFireDetected", user!!.name),
              )
              logger
                  .atInfo()
                  .log(
                      "AUTOUSERDUMP\t" +
                          EmuUtil.DATE_FORMAT.format(gameImpl.startDate) +
                          "\t" +
                          (if (aSequence < bSequence) aSequence else bSequence) +
                          "\t" +
                          game.id +
                          "\t" +
                          game.romName +
                          "\t" +
                          user.name +
                          "\t" +
                          user.socketAddress!!.address.hostAddress)
              //							logger.atFine().log("thisAction=" + EmuUtil.bytesToHex(thisAction) + "
              // actionA=" +
              // EmuUtil.bytesToHex(actionA) + " aCount=" + aCount + " actionB=" +
              // EmuUtil.bytesToHex(actionB) + " bCount=" + bCount + " aSequence=" + aSequence + "
              // aSequenceCount=" + aSequenceCount + " bSequence=" + bSequence + " bSequenceCount="
              // + bSequenceCount);
              break
            }
          }
        }
      } catch (e: Exception) {
        logger.atSevere().withCause(e).log("AutoFireScanner2 thread for $user caught exception!")
      } finally {
        synchronized(this) { running = false }
      }

      //			long et = (System.currentTimeMillis()-st);
      //			logger.atFine().log("Scanning completed in " + et + " ms");
    }

    init {
      sizeLimit = (maxDelay + 1) * minReps * 5
      buffer = Array(bufferSize) { ByteArray(sizeLimit) }
    }
  }

  companion object {
    // MAX DELAY, MIN REPEITIONS
    private val SENSITIVITY_TABLE =
        arrayOf(
            intArrayOf(0, 0),
            intArrayOf(2, 13),
            intArrayOf(3, 11),
            intArrayOf(4, 9),
            intArrayOf(5, 7),
            intArrayOf(6, 5))
    protected var executor = Executors.newCachedThreadPool()
  }

  init {
    this.sensitivity = sensitivity
  }
}
